slatedb/src/utils.rs

```diff
@@ -316,6 +316,133 @@ pub async fn timeout<T>(
     }
 }
 
+/// A simple bit-level writer that packs bits into a `[u8]`
+pub(crate) struct BitWriter {
+    buf: Vec<u8>,
+    cur: u8, // the currently assembling byte
+    n: u8,   // the number of "filled" bytes in `cur`
+}
+
+impl BitWriter {
+    pub(crate) fn new() -> Self {
+        BitWriter {
+            buf: Vec::new(),
+            cur: 0,
+            n: 0,
+        }
+    }
+
+    /// Push a single bit into the buffer
+    pub(crate) fn push(&mut self, bit: bool) {
+        if bit {
+            self.cur |= 1 << (7 - self.n);
+        }
+        self.n += 1;
+        if self.n == 8 {
+            self.flush_byte();
+        }
+    }
+
+    /// Push up to a u32 into the buffer, only considering
+    /// the first `bits` number of bits
+    pub(crate) fn push32(&mut self, value: u32, bits: u8) {
+        // writes the lowest `bits` bits from `value` into
+        // the current buffer (most significant bits first)
```

Comment:
nit: this is least significant first right? the lowest bits will be earlier in the buffer (but higher in each byte)